usr/lib/common/key.c

```diff
@@ -335,6 +351,9 @@ CK_RV key_object_validate_attribute(TEMPLATE *tmpl, CK_ATTRIBUTE *attr,
          */
         TRACE_ERROR("%s\n", ock_err(ERR_ATTRIBUTE_READ_ONLY));
         return CKR_ATTRIBUTE_READ_ONLY;
+    case CKA_IBM_PROTKEY_EXTRACTABLE:
```

Comment:
The way it is currently, CKA_IBM_PROTKEY_EXTRACTABLE can be set or modified with out any restriction. So you can use C_SetAttributeValues after object creation to change the value of CKA_IBM_PROTKEY_EXTRACTABLE. 
Is this intended? 

The PKCS#11 CKA_EXTRACTABLE attribute is restricted in how it can be modified. Per PKCS#11 the CKA_EXTRACTABLE can be modified after object creation with C_SetAttributeValues (footnote 8), but it also has footnote 12 saying: "Attribute cannot be changed once set to CK_FALSE. It becomes a read only attribute".

Do we want to also restrict CKA_IBM_PROTKEY_EXTRACTABLE that way?

Reply:
As this is an IBM internal attribute, I'm currently leaving it up to the ep11lib to decide which modification are valid. See e.g. testcase do_SetAttributeValuesPkey in aes_func.c. The ep11lib allows to change CKA_IBM_PROTKEY_EXTRACTABLE from true to false, but not vice versa.

Reply:
OK, so the EP11 library does in fact enforce the same restrictions for CKA_IBM_PROTKEY_EXTRACTABLE as for  CKA_EXTRACTABLE. 

However, since key_object_validate_attribute is a common code function, and thus the CKA_IBM_PROTKEY_EXTRACTABLE can be used with all tokens, not only EP11, I think we should enforce the restrictions in common code as well. We want that the attributes behave the same across all tokens. This won't change anything for EP11, but for the other tokens.

All you need to do is to use the following code for the `case CKA_IBM_PROTKEY_EXTRACTABLE`: block instead of just returning CKR_OK:

```
            CK_BBOOL value;

            if (attr->ulValueLen != sizeof(CK_BBOOL) || attr->pValue == NULL) {
                TRACE_ERROR("%s\n", ock_err(ERR_ATTRIBUTE_VALUE_INVALID));
                return CKR_ATTRIBUTE_VALUE_INVALID;
            }

            value = *(CK_BBOOL *) attr->pValue;
            if (mode != MODE_CREATE && mode != MODE_DERIVE &&
                 mode != MODE_KEYGEN && mode != MODE_UNWRAP &&
                 value != FALSE) {
                TRACE_ERROR("%s\n", ock_err(ERR_ATTRIBUTE_READ_ONLY));
                return CKR_ATTRIBUTE_READ_ONLY;
            }
           return CKR_OK;
```

Reply:
Done.